llvm/include/llvm/CodeGen/TargetLowering.h

```diff
@@ -1639,6 +1639,30 @@ class TargetLoweringBase {
            getCondCodeAction(CC, VT) == Custom;
   }
 
+  /// Return how a PARTIAL_REDUCE_U/SMLA node with Acc type AccVT and Input type
+  /// InputVT should be treated. Either it's legal, needs to be promoted to a
+  /// larger size, needs to be expanded to some other code sequence, or the
+  /// target has a custom expander for it.
+  LegalizeAction getPartialReduceMLAAction(EVT AccVT, EVT InputVT) const {
+    auto AccSVT = AccVT.getSimpleVT();
+    auto InputSVT = InputVT.getSimpleVT();
+    assert(AccSVT.isValid() && InputSVT.isValid() &&
+           "getPartialReduceMLAAction types aren't valid");
+    auto AccI = AccSVT.SimpleTy;
+    auto InputI = InputSVT.SimpleTy;
+    PartialReduceActionTypes TypeHash = std::make_pair(AccI, InputI);
+    if (PartialReduceMLAActions.contains(TypeHash))
+      return PartialReduceMLAActions.at(TypeHash);
```

Comment:
Two nits: 

1. `TypeHash` -> `TypePair`
2.  We can avoid doing two hash lookups by using `.lookup()` here. This will make the default action "Legal" rather than "Expand" (but we're not relying on the default anymore, we set a value for all integer scalable vector pairs, and other random type pairs don't matter). 

```suggestion
    PartialReduceActionTypes TypePair = std::make_pair(AccI, InputI);
    return PartialReduceMLAActions.lookup(TypePair);
```